### PR TITLE
Add dynamic loading for conditions and constraints

### PR DESCRIPTION
**Pull request by bjoernschroeder**

### CODE REVIEW

File: supplementary/alica_ros1/alica_tests/CMakeLists.txt [hunk 1]
```diff
@@ -62,38 +62,38 @@ if(BUILD_TESTING)
...
-        src/test/test_adjacent_plans_success.cpp
-        src/test/test_alica_tracing.cpp
-        src/test/test_blackboard.cpp
-        src/test/test_inherit_blackboard.cpp
-        src/test/test_failure_handling.cpp
-        src/test/test_alica_logger.cpp
-        src/test/test_alica_engine_worldmodel.cpp
-        src/test/test_legacy_conditions.cpp
-        src/test/test_success.cpp
+        # src/test/test_alica_init_shutdown.cpp
```
Comment (Dale-Koenig) [inline supplementary/alica_ros1/alica_tests/CMakeLists.txt]: undo commenting

**Review by Dale-Koenig — APPROVED** [review]

**Comment by bjoernschroeder** [comment]
@veerajsk We can use CRTP for constraints and conditions as well. I think we should get this one in and afterwards decide where we want to go with it in a separate PR. All projects have their own CRTP classes for alica elements with basically the same code. We could provide these in the engine, for example, engine/utils/RunnablePlan.h or something like that, so we dont have to create the same classes everywhere.

File: supplementary/alica_ros1/alica_tests/src/test/Expr/include/alica_tests/AdjacentSuccessMasterPlan.h [hunk 2]
```diff
@@ -15,17 +15,6 @@ class AdjacentSuccessMasterPlan : public DomainPlan
     AdjacentSuccessMasterPlan(PlanContext& context);
 };
 
-class PreCondition807250359520655888 : public DomainCondition
-{
-public:
-    bool evaluate(std::shared_ptr<RunningPlan> rp, const Blackboard* gb);
```
Comment (veerajsk) [inline supplementary/alica_ros1/alica_tests/src/test/Expr/include/alica_tests/AdjacentSuccessMasterPlan.h:21]: where is the implementation & export corresponding to this condition? The .pml file is not changed either so its not using some existing condition I suppose? Or is it done in some other PR

File: supplementary/alica_ros1/alica_tests/src/test/Expr/include/alica_tests/AttackPlan.h [hunk 3]
```diff
@@ -3,27 +3,27 @@
...
-public:
-    bool evaluate(std::shared_ptr<RunningPlan> rp, const Blackboard* gb);
+    void getConstraint(std::shared_ptr<ProblemDescriptor> c, std::shared_ptr<RunningPlan> rp);
 };
-class PreCondition1402489462088 : public DomainCondition
+class Constraint1402489462088 : public BasicConstraint
 {
-public:
-    bool evaluate(std::shared_ptr<RunningPlan> rp, const Blackboard* gb);
+    void getConstraint(std::shared_ptr<ProblemDescriptor> c, std::shared_ptr<RunningPlan> rp);
```
Comment (veerajsk) [inline supplementary/alica_ros1/alica_tests/src/test/Expr/include/alica_tests/AttackPlan.h:30]: don't you have to have a create function here & also export this constraint?

File: supplementary/alica_ros1/alica_tests/src/test/Expr/include/alica_tests/GoalPlan.h [hunk 4]
```diff
@@ -3,46 +3,62 @@
...
+class GoalPlanRuntimeCondition : public DomainCondition
 {
 public:
+    GoalPlanRuntimeCondition(ConditionContext& context)
+            : DomainCondition(){};
     bool evaluate(std::shared_ptr<RunningPlan> rp, const Blackboard* gb);
+    static std::unique_ptr<GoalPlanRuntimeCondition> create(ConditionContext& context) { return std::make_unique<GoalPlanRuntimeCondition>(context); };
 };
-class PreCondition1402489218027 : public DomainCondition
+class GoalPlanPostCondition : public DomainCondition
```
Comment (veerajsk) [inline supplementary/alica_ros1/alica_tests/src/test/Expr/include/alica_tests/GoalPlan.h:37]: we can get rid of DomainCondition entirely & direcly inherit here from BasicCondition

File: supplementary/alica_ros1/alica_tests/src/test/Expr/include/alica_tests/GoalPlan.h [hunk 5]
```diff
@@ -3,46 +3,62 @@
...
 {
 public:
+    GoalPlanPostCondition(ConditionContext& context)
```
Comment (veerajsk) [inline supplementary/alica_ros1/alica_tests/src/test/Expr/include/alica_tests/GoalPlan.h:40]: You can also avoid defining the constructor entirely by just ignoring the context parameter in the create function. Have a look at the CircleRuntimeCondition.h in turtlesim

File: supplementary/alica_ros1/alica_tests/src/test/Expr/include/alica_tests/GoalPlan.h (hunk 4 above)
Comment (veerajsk) [reply supplementary/alica_ros1/alica_tests/src/test/Expr/include/alica_tests/GoalPlan.h:37]: can do in follow up PR

File: supplementary/alica_ros1/alica_tests/src/test/Expr/include/alica_tests/GoalPlan.h (hunk 5 above)
Comment (veerajsk) [reply supplementary/alica_ros1/alica_tests/src/test/Expr/include/alica_tests/GoalPlan.h:40]: can do in a follow up PR

**Review by veerajsk — APPROVED** [review]

**Pull request merged**